### ISSUE TITLE
Property order in files written by Verify is machine-dependent (and I can't find a way to sort them)

### ISSUE THREAD

**Issue by bmueller84** [VerifyTests/Verify#396]
I use Verify and Verify.Xunit (11.24.0 on Framework 4.8) and have problems with inconsistent property orderings in the generated text files, which is why my tests will fail on some machines (in some sense, similar to [this bug](https://github.com/VerifyTests/Verify/issues/385)). I don't really know _why_ the ordering differs across machines, though...

Anyway, is there a known workaround for this (except for slapping json tags onto the objects to be serialized)? My initial approach was to write a custom Json contract resolver

    public class OrderedContractResolver : DefaultContractResolver
    {
        protected override IList<JsonProperty> CreateProperties(Type type, MemberSerialization memberSerialization)
            => base.CreateProperties(type, memberSerialization)
                .OrderBy(p => p.Order ?? int.MaxValue)
                .ThenBy(p => p.PropertyName)
                .ToList();
    }

and to pass it to the serializer:

    return Verifier.Verify(objectUnderTest).AddExtraSettings(s => s.ContractResolver = new OrderedContractResolver());

This kinda works but breaks a lot of things down the road. For example,

    return Verifier.Verify(objectUnderTest)
        .AddExtraSettings(s => s.ContractResolver = new OrderedContractResolver())
        .ModifySerialization(s => s.IgnoreMember<SomeType>(p => p.SomeProperty))

won't work anymore because, as far as I understand, Verify makes use of the (non-public) `CustomContractResolver` to take care of things like ignored properties. If `CustomContractResolver` was public, I might simply be able to derive my `OrderedContractResolver` from it. On the other hand, it feels like having a consistent property order should be general goal (and not something the user has to implement). Any thoughts on this?

**Comment by SimonCropp** [comment]
firstly. this is a bug and you should not need to work around it.

Can you share the model, and the different orders it produces?

Do the machines have different .net sdks installed?

**Comment by bmueller84** [comment]
Still working on a minimal reproduction (code under test has quite heavy dependencies and my initial attempt with just deleting 90% of the code failed. But I'm working on it). Both machines run .Net 4.8, but one of them is Windows 10 and the other Windows Server 2019 (shouldn't matter, I guess, just mentioning because the frameworks have [different release numbers](https://docs.microsoft.com/en-us/dotnet/framework/migration-guide/how-to-determine-which-versions-are-installed)).

But while I'm working on it: As far as I understand, you're using Newtonsoft to create the json-like text files. At the same time, the order of attributes in the created json is (by definition) undefined (that is, the order Newtonsoft spits the attributes out is an _implementation detail_ that you can't rely on). Does Verify take any measures to ensure a consistent ordering?

**Comment by SimonCropp** [comment]
> Does Verify take any measures to ensure a consistent ordering

No it doesnt. it relies on the newtonsoft order, which relies on the reflection order, which relies on the order in IL, which *should* be deterministic. 

And TBH i am not sure if it should change the order. aside from it being a significant breaking change, it is nice that it is the same order as the class

In terms of your repro. can you include the dlls from each machine. i would like to look at the IL

I could add an opt-in feature to, but i would like to determine if there is an alternative before adding a new feature i need to support.

**Comment by bmueller84** [comment]
I get your point; such a nuclear breaking change will make many users unhappy ;) In case this turns out to be something I have to live with: Would making `CustomContractResolver` public be an option (s.t. I could derive my `OrderedContractResolver` from it)?

Regarding the problem itself: Was able to dig a little deeper (still no minimal repro, unfortunately) and discovered that my initial assumption was wrong. It's not about the machine; was able to reproduce it on my machine, too. The key seems to be how I run the tests. When I run them from VS, everything is fine, but our build server is using the xunit console runner (which is .Net Framework 4.7.2). When I do the same on my machine, the tests fail, too. I still don't understand why this makes a difference (e.g., we're talking about the exact same Newtonsoft.Json.dll), but I'm gonna continue digging on Monday.

**Comment by SimonCropp** [comment]
does the same problem occur if you use a `dotnet test` command?

**Comment by bmueller84** [comment]
> does the same problem occur if you use a `dotnet test` command?

Yes, it does. And it turns out it's not about the binaries, but about the order of execution of the tests. I know it sounds strange, but here's what I tried this morning:
- Disabled test parallelization (`[assembly: CollectionBehavior(DisableTestParallelization = true)]`)
- Compiled the code once (Release configuration, if that matters)
- Ran the tests multiple times from the command line (either using the xunit console runner and `dotnet test`)
- At first, I ran the offending tests only. They _always succeeded_ with xunit console and `dotnet test`
- Now I ran the full test suite (within this one assembly) over and over again. Result is that the tests _sometimes_ fail (and the only difference in the output is the order of the attributes)
- The problem goes away if I use the `OrderedContractResolver` described above (again, with the disadvantage that I can't use all features of Verify in that case)

**Comment by SimonCropp** [comment]
can you share the model(s) for the failing test

**Comment by bmueller84** [comment]
Sorry, still not in a position to do that. Code under test is closed source, so I can't just point you to the code. At the same time, I'm still failing to extract a minimal example that reproduces the bug. I suspect either a race condition or cross-talk between tests... in any case, I'm still gathering more evidence.

Another thing I just realized: The current behavior of Verify implies that my tests will fail when I reorder the properties in a class (even without changing any functionality). Personally, I don't like this too much and would prefer to have a fixed ordering. Would you accept a corresponding pull request if I'd be able to implement this?

**Comment by bmueller84** [comment]
I "fixed" the problem on my side today. The problem was that, as far as I can tell, another test caused the properties to appear in a different order (don't ask me how, that's just what I observed and I'm still confused), and it turns out that Newtonsoft cached this wrong order _before Verify was even involved_ (Newtonsoft only does the reflection part once per type, and then retrieves the type information from a cache within `DefaultContractResolver`). I was thus able to fix the problem by ensuring that this wrong order never gets into the cache. Sorry about the hassle.

I'm gonna close this bug, but I'm still interested in the sorting discussion. I wrote some code and some tests that achieve what I'm looking for, and will open a pull request later.
